### PR TITLE
Add Language Parsing ability

### PR DESCRIPTION
**Pull request by wimpykid26**
Add Language Parsing ability

### CODE REVIEW

**Comment by wimpykid26** [comment]
@526avijitgupta Please review. Although I don't know how the merge commit came in. Its not showing in my rebase log.

**Comment by 526avijitgupta** [comment]
@wimpykid26 The linting tests are failing: https://travis-ci.org/osdc/osdc-bot/jobs/161993377
See README.md on how to run the linter locally and fix them!

**Comment by 526avijitgupta** [comment]
Also, please follow this same procedure for server.py so that the diff shows up correctly: https://github.com/osdc/osdc-bot/pull/6#issuecomment-244581311

**Comment by wimpykid26** [comment]
O
C:\Users\Dwaipayan\Desktop\bot3\osdc-bot>npm run-script lint

> osdc-bot@0.0.1 lint C:\Users\Dwaipayan\Desktop\bot3\osdc-bot
> (./node_modules/.bin/eslint ./**/*.js --quiet ; find . -path ./node_modules -p
rune -o -name '*.js' -print | xargs ./node_modules/jshint/bin/jshint) && echo 'L
int Successfull!'

'.' is not recognized as an internal or external command,
operable program or batch file.

npm ERR! Windows_NT 6.2.9200
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\
node_modules\\npm\\bin\\npm-cli.js" "run-script" "lint"
npm ERR! node v4.5.0
npm ERR! npm  v2.15.9
npm ERR! code ELIFECYCLE
npm ERR! osdc-bot@0.0.1 lint: `(./node_modules/.bin/eslint ./**/*.js --quiet ; f
ind . -path ./node_modules -prune -o -name '*.js' -print | xargs ./node_modules/
jshint/bin/jshint) && echo 'Lint Successfull!'`
npm ERR! Exit status 255
npm ERR!
npm ERR! Failed at the osdc-bot@0.0.1 lint script '(./node_modules/.bin/eslint .
/**/*.js --quiet ; find . -path ./node_modules -prune -o -name '*.js' -print | x
args ./node_modules/jshint/bin/jshint) && echo 'Lint Successfull!''.
npm ERR! This is most likely a problem with the osdc-bot package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     (./node_modules/.bin/eslint ./**/*.js --quiet ; find . -path ./node
_modules -prune -o -name '*.js' -print | xargs ./node_modules/jshint/bin/jshint)
 && echo 'Lint Successfull!'
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs osdc-bot
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!
npm ERR!     npm owner ls osdc-bot
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Dwaipayan\Desktop\bot3\osdc-bot\npm-debug.log

C:\Users\Dwaipayan\Desktop\bot3\osdc-bot>

**Review by yashmehrotra — CHANGES_REQUESTED** [review]
Please use `flake8` to lint your python code.

**Comment by 526avijitgupta** [comment]
@wimpykid26 Any updates on this ?

**Comment by wimpykid26** [comment]
Diff still not parsing correctly.

**Comment by 526avijitgupta** [comment]
@wimpykid26 Hmm. Leave off the diff issue for now. Apart from that can you address yash and my comments ?

**Comment by wimpykid26** [comment]
Yes. Will positively do this over the weekend.

**Comment by 526avijitgupta** [comment]
sgtm!

**Comment by wimpykid26** [comment]
@526avijitgupta would you please review?

**Comment by wimpykid26** [comment]
@yashmehrotra used Flake8 to lint server.py

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js:20]: No need of `|| false` in both lines.

File: server.py
Comment (526avijitgupta) [inline server.py]: Why leave newline here ?

File: server.py
Comment (526avijitgupta) [inline server.py]: Use a more descriptive name than 'general'. Something like `message_parser` ?

File: server.py
Comment (526avijitgupta) [inline server.py]: What does this regex exactly do ?

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Instead of requiring request here and then making an http request for message handling. I would suggest that you create a new method inside api.js (Named something like - parseUserMessage or something) and then simply call it here in this file!

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Incorrect indentation of this whole block.

File: server.py
Comment (526avijitgupta) [inline server.py]: `outputCommand += 'wiki'`

File: server.py
Comment (526avijitgupta) [inline server.py]: `if x[0] in ['wiki', 'wikipedia']:` should also work I guess

File: server.py
Comment (526avijitgupta) [inline server.py]: Remove brackets from every if/elif

File: server.py
Comment (526avijitgupta) [inline server.py]: Use `+=` wherever possible.

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: space after `else`

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Use a better name for this function. Also, space after comma

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Use single quotes

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: space after comma

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Suggest: Leave newline above this

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js:53]: Why remove this ?

**Review by 526avijitgupta — CHANGES_REQUESTED** [review]
@wimpykid26 Took an initial review pass.

File: server.py
Comment (wimpykid26) [reply server.py]: @526avijitgupta Breaks the sentences up into chunks containing the actual nouns (nouns verbs etc)

File: server.py
Comment (wimpykid26) [reply server.py]: @526avijitgupta flake8 suggested that. Removed it anyways.

**Comment by wimpykid26** [comment]
@526avijitgupta please review

File: server.py
Comment (526avijitgupta) [reply server.py]: @wimpykid26 If flake8 suggested that, then it should not be removed. Sorry I should've known that earlier!

File: api.js
Comment (526avijitgupta) [inline api.js]: Please avoid using underscores in JS. Thus, rename this to `getParsedMessage` or something.
FYI, we follow [Google's JS style guide](https://google.github.io/styleguide/javascriptguide.xml)

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: I think you missed my earlier comment about renaming this function to something more descriptive and clear.

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: space after ,

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: deindent whole block by 2

File: constants.js
Comment (526avijitgupta) [inline constants.js]: s/SERVER_MESSAGE_PARSER_URL/SERVER_MESSAGE_PARSER_PREFIX_URL
(Add `PREFIX`)

**Review by 526avijitgupta — CHANGES_REQUESTED** [review]
@wimpykid26 I've left a few minor comments. The PR generally looks ok. I'm still afraid about reviewing the nltk part in server.py functionality wise. @yashmehrotra Can you do that ?

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Move this else to the end of the previous line.

```
} else {
```

Btw, why is this needed  ?

File: server.py
Comment (yashmehrotra) [inline server.py]: Please make it, 

```
from flask import (
    Flask,
    request
)
```

File: server.py
Comment (yashmehrotra) [inline server.py]: @wimpykid26 Move the parsing logic to a different file

**Review by yashmehrotra — CHANGES_REQUESTED** [review]

**Comment by wimpykid26** [comment]
@526avijitgupta @yashmehrotra Please review.

**Comment by 526avijitgupta** [comment]
@wimpykid26 I can see that a few of the comments left earlier are not resolved yet. For each of the review comment - reply with "Done" if it has been resolved, otherwise reply with a reason for not following the review comment (could be because you disagree, or maybe you need clarification on a few points etc).

File: config.js.sample
Comment (526avijitgupta) [inline config.js.sample]: Don't commit the changes to this file!

File: constants.js
Comment (526avijitgupta) [inline constants.js]: Why change this ?

File: server.py
Comment (526avijitgupta) [inline server.py]: Directly `return message_chunk(command)` ?

File: server.py
Comment (526avijitgupta) [inline server.py]: This seems to extend to more than 80 chars. Therefore, suggesting to break up like this:

``` py
start_bot_comm = (
    'screen -r "{0}" -X stuff "{1}"'.format(
        screen_name, "DEPLOY=1 nodejs chatbot.js\n"))
```

File: message_chunker.py
Comment (526avijitgupta) [inline message_chunker.py:9]: Please add a docstring explaining what this method does! (See Multi-line docstrings on [here](https://www.python.org/dev/peps/pep-0257) as a reference)

File: message_chunker.py
Comment (526avijitgupta) [inline message_chunker.py]: camelCase is preferred in JS, while underscores are preferred in Python. Please use underscores for names instead of camelCase in this file as well as in server.py!

File: server.py
Comment (wimpykid26) [reply server.py]: Done

File: server.py
Comment (wimpykid26) [reply server.py]: Done

File: chatbot.js
Comment (wimpykid26) [reply chatbot.js]: Response if the input command does not fall under any category . Could be turned into an error message such as "No such command exists".

File: constants.js
Comment (wimpykid26) [reply constants.js]: Done

File: chatbot.js
Comment (wimpykid26) [reply chatbot.js]: Done

File: chatbot.js
Comment (wimpykid26) [reply chatbot.js]: renamed to 'getBotReply'

File: api.js
Comment (wimpykid26) [reply api.js]: Done.

**Review by 526avijitgupta — CHANGES_REQUESTED** [review]
@wimpykid26 Tooks another pass. I think it would be LGTM from my side once all pending review comments have been addressed!

Another thing - I'm a bit concerned about what will happen when a new functionality is added to the bot. Let's say I plan on using the thesaurus API to show the synonyms of any word that is passed as an argument! So, it would be run by using something like `@osdc-bot synonyms <word>`. Will this work directly without changes to the regular Expression that you have constructed on the backend ? If not, then how much extra effort do you think it requires ?

File: message_chunker.py
Comment (wimpykid26) [reply message_chunker.py]: I had earlier done just the same, but Flake8 suggests otherwise!

Comment (526avijitgupta) [reply message_chunker.py]: ok then.

Comment (526avijitgupta) [reply message_chunker.py]: Actually, @yashmehrotra do you think we should be using camelCase in Python ? I'm a bit surprised that flake8 is suggesting that!

**Comment by wimpykid26** [comment]
@526avijitgupta will depend on what the new command is. In this case 'synonym' is a noun and the current Regex is built to extract nouns from the sentence. Also the 'word' which will follow the command also has to be a noun/proper noun in order to be recognised and extracted by the Regex.

File: server.py
Comment (yashmehrotra) [reply server.py]: Please use a more descriptive variable name than x

File: message_chunker.py
Comment (yashmehrotra) [reply message_chunker.py]: @526avijitgupta No, flake8 doesn't suggest camelCase.

@wimpykid26 Can you send a screenshot showing flake8 suggesting camelCase

**Comment by 526avijitgupta** [comment]
@wimpykid26 Sounds good.

File: message_chunker.py
Comment (wimpykid26) [reply message_chunker.py]: Apologies. You are right. Removing camelCase

File: message_chunker.py
Comment (wimpykid26) [reply message_chunker.py:9]: Done

File: server.py
Comment (wimpykid26) [reply server.py]: Done.

File: server.py
Comment (wimpykid26) [reply server.py]: Done

File: constants.js
Comment (wimpykid26) [reply constants.js]: For testing purposes. Changed it back now. Apologies !

**Comment by wimpykid26** [comment]
@526avijitgupta @yashmehrotra Please review.

File: chatbot.js
Comment (526avijitgupta) [reply chatbot.js]: @wimpykid26 Looks like you missed this one!

Comment (wimpykid26) [reply chatbot.js]: Done

**Comment by 526avijitgupta** [comment]
@wimpykid26 Final things before merge:
- Please add nltk to requirements.txt in the same format as the already added libraries! (to know the version of nltk you are running locally, run `pip freeze`).
- Please add `*.pyc` to .gitignore to ignore files generated on executing python scripts!
-  Update your branch with master and fix merge conflicts:

``` bash
git fetch upstream
git merge upstream/master
# fix merge conflicts in constants.js
git add constats.js
git commit -m "update branch and fix merge conflicts"
git push origin master
```

The current code looks good to me!

Also, I tried this out by checking out your branch locally, and it works great 😄 !!

**Comment by wimpykid26** [comment]
@526avijitgupta Done. Please review.

File: constants.js
Comment (526avijitgupta) [inline constants.js]: Replace this by `JOKES_API_INFO`

File: constants.js
Comment (526avijitgupta) [inline constants.js]: Remove this.

File: chatbot.js
Comment (526avijitgupta) [inline chatbot.js]: Indentation seems to be 3-spaced! Please make it 2 spaced.

File: requirements.txt~
Comment (526avijitgupta) [inline requirements.txt~]: Don't commit this file.

**Review by 526avijitgupta — CHANGES_REQUESTED** [review]
@wimpykid26 Minor comments

File: constants.js
Comment (526avijitgupta) [reply constants.js]: *Remove this as well.

File: constants.js
Comment (wimpykid26) [reply constants.js]: The JOKES_API_URL right?

Comment (526avijitgupta) [reply constants.js]: Yep. It's no longer needed since it has now been substituted by `JOKES_API_INFO`!

File: requirements.txt~
Comment (wimpykid26) [reply requirements.txt~]: It is ignoring the file.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

```
deleted:    requirements.txt~
```

no changes added to commit (use "git add" and/or "git commit -a")
dwaipayan@dwaipayan-HP-ENVY-15-Notebook-PC:~/Bot/osdc-bot$ git add .
warning: You ran 'git add' with neither '-A (--all)' or '--ignore-removal',
whose behaviour will change in Git 2.0 with respect to paths you removed.
Paths like 'requirements.txt~' that are
removed from your working tree are ignored with this version of Git.
- 'git add --ignore-removal <pathspec>', which is the current default,
  ignores paths you removed from your working tree.
- 'git add --all <pathspec>' will let you also record the removals.

Run 'git status' to check the paths you removed from your working tree.

Comment (526avijitgupta) [reply requirements.txt~]: Use `git rm <file_name>` . Then commit and push!

Comment (wimpykid26) [reply requirements.txt~]: Done.

**Review by 526avijitgupta — APPROVED** [review]
LGTM

**Comment by yashmehrotra** [comment]
@wimpykid26 LGTM

@526avijitgupta Can you test this locally, the functionality and all. If it is good, merge it.
Also, we need to install pip dependencies manually on `bludhaven` (keep that in mind)

**Review by yashmehrotra — APPROVED** [review]

**Comment by 526avijitgupta** [comment]
@wimpykid26 Works good. Merging!

**Pull request merged**
